webapi/src/main/scala/org/knora/webapi/messages/admin/responder/permissionsmessages/PermissionsMessagesUtilADM.scala

```diff
@@ -88,4 +105,82 @@ object PermissionsMessagesUtilADM {
 
     CacheUtil.remove(PermissionsCacheName, key)
   }
+
+  def validateDOAPHasPermissions(hasPermissions: Set[PermissionADM]) = {
+
+    hasPermissions.foreach { permission =>
+      if (permission.additionalInformation.isEmpty) {
+        throw BadRequestException(s"additionalInformation of a default object access permission type cannot be empty.")
+      }
+      if (permission.name.nonEmpty && !EntityPermissionAbbreviations.contains(permission.name))
+        throw BadRequestException(
+          s"Invalid value for name parameter of hasPermissions: ${permission.name}, it should be one of " +
+            s"${EntityPermissionAbbreviations.toString}")
+      if (permission.permissionCode.nonEmpty) {
+        val code = permission.permissionCode.get
+        if (!PermissionTypeAndCodes.values.toSet.contains(code)) {
+          throw BadRequestException(
+            s"Invalid value for permissionCode parameter of hasPermissions: $code, it should be one of " +
+              s"${PermissionTypeAndCodes.values.toString}")
+        }
+      }
+      if (permission.permissionCode.isEmpty && permission.name.isEmpty) {
+        throw BadRequestException(
+          s"One of permission code or permission name must be provided for a default object access permission.")
+      }
+      if (permission.permissionCode.nonEmpty && permission.name.nonEmpty) {
+        val code = permission.permissionCode.get
+        if (PermissionTypeAndCodes(permission.name) != code) {
+          throw BadRequestException(
+            s"Given permission code $code and permission name ${permission.name} are not consistent.")
+        }
+      }
+    }
+  }
+
+  /**
+    * For administrative permission we only need the name parameter of each PermissionADM given in hasPermissions collection.
+    * This method, validates the content of hasPermissions collection by only keeping the values of name params.
+    */
+  def verifyHasPermissionsAP(hasPermissions: Set[PermissionADM]): Set[PermissionADM] = {
+    val updatedPermissions = hasPermissions.map { permission =>
+      if (!AdministrativePermissionAbbreviations.contains(permission.name))
+        throw BadRequestException(
+          s"Invalid value for name parameter of hasPermissions: ${permission.name}, it should be one of " +
+            s"${AdministrativePermissionAbbreviations.toString}")
+      PermissionADM(
+        name = permission.name,
+        additionalInformation = None,
+        permissionCode = None
+      )
+    }
+    updatedPermissions
+  }
+
+  /**
+    * For default object access permission, we need to make sure that the value given for the projectCode matches the value of name parameter.
+    * This method, validates the content of hasPermissions collection by verifying that both projectCode and name indicate the same type of permission.
```

Comment:
what is meant by projectCode?

Reply:
hehe that was a permissioncode, sorry for the typo.